API/controllers/taskController.js

```diff
@@ -131,5 +136,70 @@ export const taskController = {
         message: err.message
       })
     }
+  },
+
+  notifyUser: async (req, res) => {
+    try {
+      const now = new Date();
+      const tasks = await Task.find();
+
+      tasks.forEach(async (task) => {
+        const dueDate = new Date(task.dueDate);
+        const timeLeft = dueDate - now;
+
+        const timeBeforeNotification = (24 * 60 * 60 * 1000);  // alert 24hrs before due date
+
+        if (timeLeft <= timeBeforeNotification && task.status !== 'completed' && !task.notified) {
+          const user = await req.findById(req.params.id);
+          if (user && user.email) {
+            sendNotification(user.email, task.title, timeLeft);
+            task.notified = true;
+            await task.save({
+              validateBeforeSave: true
+            })
+      
+            res.status(200).json({
+              status: 'success',
+              data: {
+                task
+              }
+            })
+          }
+        }
+      });
+    } catch (err) {
+      res.status(400).json({
+        status: 'fail',
+        message: err.message
+      })
+    }
   }
+
 }
+
+const sendNotification = (email, taskTitle, timeLeft) => {
+  const transporter = nodemailer.createTransport({
+    service: 'Gmail',
+    auth: {
+      user: process.env.NOTIF_EMAIL,
+      pass: process.env.NOTIF_PASS, 
+    },
+  });
+
+  const mailOptions = {
+    from: 'em.task.notify@gmail.com',
+    to: email,
+    subject: 'Task Reminder',
+    text: `Reminder! Your task "${taskTitle}" is due in ${Math.round(timeLeft / 3600000)} hours.`,
+  };
+
+  transporter.sendMail(mailOptions, (err, info) => {
+    if (err) {
+      console.log('Error sending email:', err.message);
+    } else {
+      console.log('Email sent:', info.response);
+    }
+  });
+};
+
+setInterval(taskController.notifyUser, 60000);  // 1min interval
```

Comment:
I agree, Ideally this will be able to work for users who want "notify me __ hours before its due, __ days before due etc." the function is there for flexibility